compiler/hash-typecheck/src/exhaustiveness/fields.rs

```diff
@@ -0,0 +1,145 @@
+//! Data structure that represents the `fields` of a
+//! [DeconstructedPat](super::deconstruct::DeconstructedPat). The [Fields] data
+//! structure is an inner [Vec] of [DeconstructedPatId]s. It has some useful
+//! creation methods for when
+//! [DeconstructedPat](super::deconstruct::DeconstructedPat)s need to be created
+//! from a provided [PatCtx]. [FieldOps] defines methods that operate on
+//! [Fields] with the typechecker context available for reading and creating
+//! [DeconstructedPat](super::deconstruct::DeconstructedPat)s.
+use itertools::Itertools;
+
+use crate::{
+    diagnostics::macros::tc_panic,
+    exhaustiveness::PatCtx,
+    ops::AccessToOps,
+    storage::{
+        primitives::{
+            ConstructorId, DeconstructedPatId, Level1Term, NominalDef, StructFields, Term, TermId,
+            TupleTy,
+        },
+        AccessToStorage, StorageRef,
+    },
+};
+
+use super::{construct::Constructor, AccessToUsefulnessOps};
+
+/// Representation of the `fields` that are stored by
+/// [DeconstructedPat](super::deconstruct::DeconstructedPat) which are nested
+/// patterns.
+#[derive(Debug, Clone)]
+pub struct Fields {
+    /// Vector of the inner ids stored by the [Fields]
+    pub fields: Vec<DeconstructedPatId>,
+}
```

Comment:
We can potentially inline store this in a store to avoid the overhead of having all these Vecs. Will tackle this later.